### PR TITLE
Use ThrowHelper methods in more places

### PR DESCRIPTION
**Pull request by turbedi**
- Might help with https://github.com/ppy/osu-framework/issues/3470

ObjectDisposedException.ThrowIf(), ArgumentException.ThrowIfNullOrEmpty() and the ArgumentOutOfRangeException methods have been introduced in .NET 7/8

### CODE REVIEW

File: osu.Framework/Bindables/BindableNumber.cs [hunk 1]
```diff
@@ -42,8 +42,7 @@ public T Precision
                 if (precision.Equals(value))
                     return;
 
-                if (value.CompareTo(default) <= 0)
-                    throw new ArgumentOutOfRangeException(nameof(Precision), value, "Must be greater than 0.");
+                ArgumentOutOfRangeException.ThrowIfNegativeOrZero(value.CompareTo(default));
```
Comment (smoogipoo) [inline osu.Framework/Bindables/BindableNumber.cs:45]: Can this be `ThrowIfNegativeOrZero(value)`?

Comment (turbedi) [reply osu.Framework/Bindables/BindableNumber.cs:45]: Yes should be fine. Will change it.

File: osu.Framework/Graphics/Containers/GridContainer.cs [hunk 2]
```diff
@@ -420,11 +420,8 @@ public class Dimension
         /// <param name="maxSize">The maximum size of this row or column.</param>
         public Dimension(GridSizeMode mode = GridSizeMode.Distributed, float size = 0, float minSize = 0, float maxSize = float.MaxValue)
         {
-            if (minSize < 0)
-                throw new ArgumentOutOfRangeException(nameof(minSize), "Must be greater than 0.");
-
-            if (minSize > maxSize)
-                throw new ArgumentOutOfRangeException(nameof(minSize), $"Must be less than {nameof(maxSize)}.");
+            ArgumentOutOfRangeException.ThrowIfNegative(minSize);
+            ArgumentOutOfRangeException.ThrowIfGreaterThan(minSize, maxSize);
```
Comment (Susko3) [inline osu.Framework/Graphics/Containers/GridContainer.cs:424]: Just noting that `maxSize` no will longer appear in the error message.

master:
```cs
System.ArgumentOutOfRangeException : Must be less than maxSize. (Parameter 'minSize')
```

This PR:
```cs
System.ArgumentOutOfRangeException : minSize ('20') must be less than or equal to '10'. (Parameter 'minSize')
Actual value was 20.
```

File: osu.Framework/Testing/Drawables/TestGroupButton.cs [hunk 3]
```diff
@@ -45,8 +45,7 @@ public TestGroupButton(Action<Type> loadTest, TestGroup group)
         {
             var tests = group.TestTypes;
 
-            if (tests.Length == 0)
-                throw new ArgumentOutOfRangeException(nameof(group), tests.Length, "Type array must not be empty!");
+            ArgumentOutOfRangeException.ThrowIfZero(tests.Length);
```
Comment (Susko3) [inline osu.Framework/Testing/Drawables/TestGroupButton.cs:48]: Should be reverted, `tests.Length` is not an argument of this method. The error message doesn't make much sense.
```cs
0 ('tests.Length') must be a non-zero value.
```

File: osu.Framework/Utils/IncrementalBSplineBuilder.cs [hunk 4]
```diff
@@ -140,8 +138,7 @@ public float CornerThreshold
             get => cornerThreshold;
             set
             {
-                if (cornerThreshold < 0)
-                    throw new ArgumentOutOfRangeException(nameof(value), "CornerThreshold must not be negative.");
+                ArgumentOutOfRangeException.ThrowIfNegative(cornerThreshold);
```
Comment (Susko3) [inline osu.Framework/Utils/IncrementalBSplineBuilder.cs:141]: Wait this is actually wrong on master, it's checking the old `cornerThreshold` and not the newly-set `value`.
Should be changed to

```cs
ArgumentOutOfRangeException.ThrowIfNegative(value);
```

File: osu.Framework/Utils/RNG.cs [hunk 5]
```diff
@@ -61,8 +60,7 @@ public static double NextDouble(double maxValue)
         /// <returns>A double-precision floating point number in the range [minValue,maxValue).</returns>
         public static double NextDouble(double minValue, double maxValue)
         {
-            if (minValue > maxValue)
-                throw new ArgumentOutOfRangeException(nameof(minValue), "The given minimum value must be less than or equal to the given maximum value.");
+            ArgumentOutOfRangeException.ThrowIfGreaterThan(minValue, maxValue);
```
Comment (Susko3) [inline osu.Framework/Utils/RNG.cs:63]: See other comment about `maxValue`/`maxSize` no longer appearing in the error message. Same applies for the other usage of `ThrowIfGreaterThan()` in this file.

File: osu.Framework/Utils/IncrementalBSplineBuilder.cs [hunk 6]
```diff
@@ -117,8 +116,7 @@ public float Tolerance
             get => tolerance;
             set
             {
-                if (tolerance < 0)
-                    throw new ArgumentOutOfRangeException(nameof(value), "Tolerance must not be negative.");
+                ArgumentOutOfRangeException.ThrowIfNegative(tolerance);
```
Comment (Susko3) [inline osu.Framework/Utils/IncrementalBSplineBuilder.cs:119]: Should be:

```cs
ArgumentOutOfRangeException.ThrowIfNegative(value);
```

**Review by Susko3 — CHANGES_REQUESTED** [review]
Seems ok but some error messages have changed for the worse.

File: osu.Framework/Bindables/BindableNumber.cs (hunk 1 above)
Comment (Susko3) [reply osu.Framework/Bindables/BindableNumber.cs:45]: Having this done via `CompareTo()` results in the actual value being misrepresented. Eg. setting a `BindableFloat.Precision` to `-5f` will show `-1` as the actual value.

```cs
System.ArgumentOutOfRangeException : value.CompareTo(default) ('-1') must be a non-negative and non-zero value. (Parameter 'value.CompareTo(default)')
Actual value was -1.
```

This is better handled via generic math (could just call `ArgumentOutOfRangeException.ThrowIfNegativeOrZero(value)`). But that is a breaking change and should not be done in this PR, so best to revert for now.

**Comment by turbedi** [comment]
I've addressed some of the comments. Should I also revert all usages of ThrowIfGreaterThan() to retain the max value in the error message, or is it fine like that?

**Comment by smoogipoo** [comment]
The other path I can suggest is to make an `ExceptionUtils.ValidateRange()`. Otherwise, I'm _weakly_ okay with the change in messaging but I'd want other reviews.

**Comment by frenzibyte** [comment]
I personally don't feel the exception message matters that much. Sure it might be a little bit cryptic, but the "must be greater than"/"must be lower than" is already giving it out I think.

**Comment by bdach** [comment]
My view on the `ThrowIfGreaterThan()` thing is that I'd be fine with the max value not being printed *only* if it's a constant. So in cases where the max value isn't statically known just by looking at the direct call site it should print the max in the exception message.

In favour of revert.

File: osu.Framework/Bindables/BindableNumber.cs (hunk 1 above)
Comment (turbedi) [reply osu.Framework/Bindables/BindableNumber.cs]: Done

**Comment by turbedi** [comment]
All comments have been addressed.

**Review by peppy — APPROVED** [review]

**Review by bdach — APPROVED** [review]

**Pull request merged**